### PR TITLE
Fixed #26165 -- Added some FAQs about CSRF protection.

### PR DESCRIPTION
**Pull request by timgraham**
https://code.djangoproject.com/ticket/26165

### CODE REVIEW

File: docs/faq/csrf.txt [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+---------------------------------------
+
+No, it is not. Django CSRF security system provides an implementation that allows you to have CSRF protection without linking it to a session. The way the CSRF token works is simple. Every form will have a CSRF token which matches the CSRF cookies. While processing the submitted form, the server will make sure that the submitted CSRF token matches the cookie value. This is a server side check but it is not against any stored server-side value. Since the remote attacker would not be able to read or set arbitrary cookies on your domain, it protects the user.
+
+Why can't we keep a new token for each request ?
+------------------------------------------------
+
+Generating a new token for each request is problematic from a UI perspective because it invalidates all previous forms. Most users would be very unhappy to find that opening a new tab on your site has invalidated the form they had just spent time filling out in the other tab or that a form they accessed via the back button could not be filled out.
+
+How can I make Django's CSRF protection work to its full extent ?
```
Comment (carljm) [inline docs/faq/csrf.txt]: Spurious space before question mark

File: docs/faq/csrf.txt [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+------------------------------------------------------------------
+
+Make sure of the following:
+
+* Use HTTPS.Redirect to the encrypted version for all unencrypted requests. If you don't do this,
```
Comment (carljm) [inline docs/faq/csrf.txt]: Missing space after period

File: docs/faq/csrf.txt [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  CSRF protection won't be able to protect you from a man-in-the-middle attack.
+
+* Use HSTS. Set it for several months. Use "includeSubDomains". This means that no matter what your
+  users type, and no matter what the man-in-the-middle does, your users will always access your site securely if they've been there at least once before.
+
+* Validate the HOST header in your httpd. Don't allow arbitrary requests to fall through to Django.
+  Serve your site only for the appropriate domain. See the most recent security advisory for more information about this.
```
Comment (carljm) [inline docs/faq/csrf.txt]: "most recent security advisory" doesn't work as a permanent wording for the docs. Need a specific date (and maybe link).

File: docs/faq/csrf.txt [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+Make sure of the following:
+
+* Use HTTPS.Redirect to the encrypted version for all unencrypted requests. If you don't do this,
+  CSRF protection won't be able to protect you from a man-in-the-middle attack.
+
+* Use HSTS. Set it for several months. Use "includeSubDomains". This means that no matter what your
+  users type, and no matter what the man-in-the-middle does, your users will always access your site securely if they've been there at least once before.
+
+* Validate the HOST header in your httpd. Don't allow arbitrary requests to fall through to Django.
```
Comment (carljm) [inline docs/faq/csrf.txt]: Does `httpd` work as a generic term for "front-end webserver or proxy"? Reads kind of Apache-centric to me.

File: docs/faq/csrf.txt [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  Serve your site only for the appropriate domain. See the most recent security advisory for more information about this.
+
+* Ensure there are no XSS vulnerabilities on your website.
```
Comment (carljm) [inline docs/faq/csrf.txt]: Any reason not to spell out "cross-site scripting" here? Not everyone knows the acronyms.

File: docs/faq/csrf.txt [hunk 6]
```diff
@@ -0,0 +1,30 @@
+FAQ: CSRF Protection
+=====================
+
+Is Django's CSRF protection vulnerable?
+---------------------------------------
+
+No, it is not. Django CSRF security system provides an implementation that allows you to have CSRF protection without linking it to a session. The way the CSRF token works is simple. Every form will have a CSRF token which matches the CSRF cookies. While processing the submitted form, the server will make sure that the submitted CSRF token matches the cookie value. This is a server side check but it is not against any stored server-side value. Since the remote attacker would not be able to read or set arbitrary cookies on your domain, it protects the user.
+
+Why can't we keep a new token for each request ?
```
Comment (carljm) [inline docs/faq/csrf.txt]: Spurious space before question mark

**Comment by acemaster** [comment]
@carljm I will correct those mistakes and add a link to security advisory.

**Comment by carljm** [comment]
@acemaster No need for apology, thanks for the PR! It may be that XSS is common enough that everyone knows it, I dunno. Doesn't hurt to spell it out, though.

@timgraham will likely have other review comments here.

File: docs/spelling_wordlist [hunk 7]
```diff
@@ -979,3 +979,5 @@ xxxxx
 yesno
 Zealand
 Zope
+includeSubDomains
```
Comment (timgraham) [inline docs/spelling_wordlist]: These terms should be surround by doubebackticks instead of added to the word list.

**Comment by timgraham** [comment]
Please wrap lines at 79 characters and link to the FAQ from docs/ref/csrf.txt. I'll give a detailed review later.

**Comment by acemaster** [comment]
@timgraham @carljm Hi I have fixed the spaces and added the relevant links. Anything else I have to add ?

**Comment by timgraham** [comment]
You need to wrap lines at 79 characters instead of putting each sentence on a single line.

File: docs/spelling_wordlist [hunk 8]
```diff
@@ -978,4 +978,4 @@ XSS
 xxxxx
 yesno
 Zealand
-Zope
+Zope
```
Comment (timgraham) [inline docs/spelling_wordlist]: this file shouldn't appear as changed

**Comment by acemaster** [comment]
@timgraham sorry about that. I have wrapped the lines at 79 characters. Anything else or any other improvements I should look into ?

**Comment by acemaster** [comment]
@timgraham Any reason why the windows and default tests are not passing in CI ? I looked through the windows test console output and I found the test was failing at 
```
pull-requests-windows » sqlite3,windows,Python27 completed with result FAILURE
```

**Comment by timgraham** [comment]
Don't worry about it, a documentation change like this doesn't affect that.

**Comment by timgraham** [comment]
Could you incorporate these edits? http://dpaste.com/0VEVVX8
Thanks!

**Comment by acemaster** [comment]
@timgraham No problem I will make these changes and push by today

**Comment by timgraham** [comment]
If you don't know about it, you can download the raw version of the patch and apply it with `patch -p1 -i 109H9SA.txt`

**Comment by acemaster** [comment]
@timgraham really sorry for the delay. I have my college exams going on. I will send a pull request by Friday. Thanks for the help

**Comment by hobarrera** [comment]
Commit summary should read "Added a FAQ", not "Added a**n** FAQ".

**Comment by timgraham** [comment]
Updated in https://github.com/django/django/pull/6190

**Pull request closed without merging**